Add diagonal movement

Gold872 opened the pull request.

Gold872 committed:
Initial test of diagonal movement

Levi-Lesches commented on lib/src/interfaces/a_star.dart:
```diff
@@ -54,13 +58,88 @@ class AutonomyAStarState extends AStarState<AutonomyAStarState> {
+    final isQuarterTurn = state.direction == DriveDirection.forwardLeft || state.direction == DriveDirection.forwardRight;
+
+    // Forward drive across the perpendicular axis
+    if (!isTurn && state.orientation.angle.abs() % 90 == 0) {
```
Can you give `state.orientation.angle.abs() % 90 == 0` a descriptive variable name?

Levi-Lesches commented on lib/src/interfaces/a_star.dart:
```diff
@@ -54,13 +58,88 @@ class AutonomyAStarState extends AStarState<AutonomyAStarState> {
+      return false;
+    }
+
+    DriveOrientation orientation1;
+    DriveOrientation orientation2;
```
What do these two represent?

Gold872 replied:
The 2 different perpendicular rotations it will check for when seeing if there's obstacles between the rover

Levi-Lesches replied:
Okay I think I get it, just elaborate a bit more in a full-on doc comment `///` or in-function comment `//`. Doc comment if the caller needs to know, regular comment otherwise. 

One trick I use for complicated logic is explicitly naming a few cases in the comments, then referencing them in the code, like this:

```dart
// There are a few cases this function handles: 
//
// 1. The first thing occurs
// 2. The second thing occurs
// 3. Maybe another thing

if (...) {  // case 1
  // ... 
} else if (...) {  // case 2
  // ...
} 

return null;  // case 3

Levi-Lesches commented on lib/src/rover/orchestrator.dart:
```diff
@@ -39,11 +39,13 @@ class RoverOrchestrator extends OrchestratorInterface with ValueReporter {
     collection.logger.debug("Currently at ${collection.gps.coordinates.prettyPrint()}");
     collection.drive.setLedStrip(ProtoColor.RED);
+    collection.detector.findObstacles();
+    // await collection.drive.faceNorth();
```
You'll have to either add this back in, or change the A* to take into account the initial orientation of the rover.

Levi-Lesches commented on lib/src/rover/orchestrator.dart:
```diff
@@ -39,11 +39,13 @@ class RoverOrchestrator extends OrchestratorInterface with ValueReporter {
       collection.logger.debug("Finding a path");
       currentState = AutonomyState.PATHING;
-      await collection.drive.faceNorth();
+      // await collection.drive.faceNorth();
```
Yeah this can be deleted with the above addition

Levi-Lesches commented on lib/src/rover/orchestrator.dart:
```diff
@@ -61,13 +63,13 @@ class RoverOrchestrator extends OrchestratorInterface with ValueReporter {
       }
       currentState = AutonomyState.DRIVING;
-      var count = 0;
+      // var count = 0;
```
This was good because it catches small drifts in the GPS and IMU, allowing us to be more lenient about errors in everything. Maybe let's keep it and try removing it later when we do more physical tests

Gold872 replied:
Maybe instead of replanning after a set amount of steps, have it check after each step is completed if it has drifted too far and replan?

Levi-Lesches replied:
Planning is very cheap -- not noticeable during operation. If there is no drift, the new plan == the old plan. So this is pretty much just as efficient as that, without the extra logic

Gold872 replied:
Doesn't replanning stop the rover suddenly? I'd rather not replan if it will cause a visible hiccup

Levi-Lesches replied:
Not if we move `faceNorth` outside the while loop, Then there is no stop command and the re-planning will happen while the rover is still executing/finishing the last command. Technically there is a very slight delay but we're still unsure exactly what accuracy we will end up having. Keep in mind that we still don't have RTK so our GPS will have significant error.

Levi-Lesches commented on lib/src/interfaces/a_star.dart:
```diff
@@ -54,13 +58,88 @@ class AutonomyAStarState extends AStarState<AutonomyAStarState> {
     direction: direction,
     goal: goal, 
-    depth: direction == DriveDirection.forward ? depth + 1 : depth + 2,
+    depth: (direction == DriveDirection.forward)
+        ? depth + 1
+        : (direction == DriveDirection.forwardLeft || direction == DriveDirection.forwardRight)
+            ? depth + sqrt2
+            : depth + 2,
```
The fact that this works is very good news because it's very reasonable :+1:

Gold872 committed:
Added turning logic to sensor and timed drive

Gold872 committed:
More docs

Gold872 committed:
Added network detector

Gold872 committed:
Refactored movement units

Bing-Rover committed:
Added tank autonomy file

Bing-Rover committed:
Tested on tank 11/25

Gold872 committed:
Fixed all occurances of toGps()

Gold872 committed:
Fixed toGps() in task.dart

Gold872 committed:
Specified networking library version

Gold872 committed:
Fixed isNear method

Gold872 committed:
Refactor and Reorganization (#7)

Co-authored-by: Binghamton Rover <rover@binghamton.edu>
Co-authored-by: Levi Lesches <levilesches@gmail.com>

Levi-Lesches commented on lib/src/utils/gps_utils.dart:
```diff
@@ -0,0 +1,73 @@
+
+  String prettyPrint() => toProto3Json().toString();
+
+  GpsCoordinates goForward(CardinalDirection orientation) => this + switch(orientation) {
+    CardinalDirection.north => GpsUtils.north,
+    CardinalDirection.south => GpsUtils.south,
+    CardinalDirection.west => GpsUtils.west,
+    CardinalDirection.east => GpsUtils.east,
+    CardinalDirection.northEast => GpsUtils.northEast,
+    CardinalDirection.northWest => GpsUtils.northWest,
+    CardinalDirection.southEast => GpsUtils.southEast,
+    CardinalDirection.southWest => GpsUtils.southWest,
+  };
```
We can make this a field on `CardinalDirection`, so that this becomes: 
```dart
GpsCoordinates goForward(CardinalDirection direction) => this + direction.gpsOffset;
```

Levi-Lesches commented on lib/src/utils/imu_utils.dart:
```diff
@@ -0,0 +1,29 @@
+
+extension OrientationUtils on Orientation {
+  static const double epsilon = 3.5;
+  static const double orientationEpsilon = 10;
```
```suggestion
```

Seems to be unused

Gold872 committed:
Corrected directions

Gold872 committed:
Initial work on fixing unit tests

Gold872 committed:
Account for move length in heuristic

Gold872 committed:
Simplified heuristic calculation (again)

Gold872 committed:
Fix path following not stopping after restart

Gold872 committed:
Fixed heuristic math to use only distance

Gold872 committed:
Add more path tests

Gold872 committed:
Fixed excessive turns (no idea why this fixes it but whatever)

Gold872 committed:
fix merge conflicts

Gold872 committed:
Fixed sensor drive

Gold872 committed:
Fixed heuristic and duplicate hashing

Bing-Rover committed:
Initial testing on rover

Gold872 committed:
Added path optimizer and more unit tests

Bing-Rover committed:
Additional testing on rover

Bing-Rover committed:
Works on rover!

Gold872 committed:
Fixes from testing
- Clamp all angles from -180 to 180
- Face the correct direction before driving forward

Tested with sim GPS and real IMU*

Gold872 committed:
Began documenting

Gold872 committed:
Use sim imu in task

Gold872 committed:
Use meters for goForward()

Gold872 committed:
Use meters for distanceTo()

Gold872 committed:
Moved things into constants

Gold872 committed:
More safety replanning features:
- Replan if the rover is too far from the path
- Add a maximum time for the rover to drive forward 1 step

Gold872 committed:
Replan paths if timeout is hit

Gold872 committed:
GPSUtil cleanup
- The directions don't need to be getters anymore

Gold872 committed:
Improved heuristic and sharpen intermediate drive tolerance

Gold872 committed:
Point to next waypoint before driving (only with RTK)

Gold872 committed:
Resolved NASTY merge conflicts

Gold872 committed:
Remove unused imports

Gold872 committed:
Update CI dependency overrides

Gold872 committed:
Return method early when timing out

Gold872 committed:
Fix hasValue

Gold872 committed:
Bump mandatory burt_network version
